### PR TITLE
Neo database

### PR DESCRIPTION
**Pull request by Christian-B**
combines work from:
https://github.com/SpiNNakerManchester/SpiNNFrontEndCommon/pull/997
https://github.com/SpiNNakerManchester/SpiNNFrontEndCommon/pull/998
https://github.com/SpiNNakerManchester/SpiNNFrontEndCommon/pull/999

View tracking the reset number

BufferManager instead holding a database it uses the with pattern

On reset the old sqlite3 is no longer deleted

BaseDatabase class created which will hold code shared with the ProvenanceReader/ProvenanceWriter when they later share the same sqlite3 file

splits off the GlobalProvenance into a separate Class and separate sqlite3 file
This includes version timer and log provenance all of which apply over multiple resets.
This sqlite3 file is therefor next to and not in the run_ directories

Must be done at the same time as:
https://github.com/SpiNNakerManchester/JavaSpiNNaker/pull/682
https://github.com/SpiNNakerManchester/sPyNNaker/pull/1269
https://github.com/SpiNNakerManchester/SpiNNakerGraphFrontEnd/pull/229
https://github.com/SpiNNakerManchester/SpiNNGym/pull/60

tested by
https://github.com/SpiNNakerManchester/IntegrationTests/pull/173

### CODE REVIEW

File: spinn_front_end_common/utilities/report_functions/bit_field_compressor_report.py [hunk 1]
```diff
@@ -56,11 +58,15 @@ def bitfield_compressor_report():
     file_name = os.path.join(FecDataView.get_run_dir_path(), _FILE_NAME)
     try:
         with open(file_name, "w", encoding="utf-8") as f:
-            return _write_report(f)
+            _write_report(f)
     except IOError:
         logger.exception("Generate_placement_reports: Can't open file"
                          " {} for writing.", _FILE_NAME)
-        return None
+        return
```
Comment (rowleya) [inline spinn_front_end_common/utilities/report_functions/bit_field_compressor_report.py:65]: Is return needed here?

File: spinn_front_end_common/utilities/report_functions/bit_field_compressor_report.py [hunk 2]
```diff
@@ -56,11 +58,15 @@ def bitfield_compressor_report():
...
+    except NoProvenanceDatabaseException:
+        logger.exception(
+            "No proveance found to write bitfield_compressor_report")
+        return
```
Comment (rowleya) [inline spinn_front_end_common/utilities/report_functions/bit_field_compressor_report.py:69]: Probably not needed here either?

**Review by rowleya — CHANGES_REQUESTED** [review]
Minor changes requested.  I didn't look in detail at the database queries, but assume these are working by tests!

File: spinn_front_end_common/data/fec_data_view.py [hunk 3]
```diff
@@ -442,6 +444,50 @@ def has_time_scale_factor(cls):
...
+            If the run_number is currently unavailable
+        """
+        if cls.__fec_data._reset_number is None:
+            raise cls._exception("run_number")
+        return cls.__fec_data._reset_number
+
+    @classmethod
+    def get_reset_str(cls):
+        """
+        Get the number of times a reset has happene as a string. Zero as ""
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spinn_front_end_common/data/fec_data_view.py:472]: It was correct as zero is an empty string but reworded to make it easier to understand

File: spinn_front_end_common/interface/buffer_management/buffer_manager.py [hunk 4]
```diff
@@ -82,8 +82,17 @@ class BufferManager(object):
...
+        "_thread_lock_buffer_out",
+
+        # Lock to avoid multiple messages being processed at the same time
+        "_thread_lock_buffer_in",
+
+        # bool flag
+        "_finished",
+
+        # listener port
+        "_listener_port",
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spinn_front_end_common/interface/buffer_management/buffer_manager.py:95]: Good catch the leaked back in during a merge-

File: spinn_front_end_common/utilities/report_functions/bit_field_compressor_report.py (hunk 2 above)
Comment (Christian-B) [reply spinn_front_end_common/utilities/report_functions/bit_field_compressor_report.py:69]: NoProvenanceDatabaseException is never raised so whole Exception classed nuked.

File: spinn_front_end_common/utilities/report_functions/bit_field_compressor_report.py (hunk 1 above)
Comment (Christian-B) [reply spinn_front_end_common/utilities/report_functions/bit_field_compressor_report.py:65]: nope removed

**Review by rowleya — APPROVED** [review]

**Pull request merged**